### PR TITLE
Refactoring recurrence configurations

### PR DESCRIPTION
**Pull request by GabrielDeveloper**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| **Issue**         | https://mundipagg.atlassian.net/browse/MOD-933
| **What?**         | Refactoring recurrence configurations
| **Why?**          | To allow the cart conflicts configuration
| **How?**          | Changing the recurrence config object and added more configurations

### CODE REVIEW

**Comment by wallaceSF** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 4
           

Coverage decreased per file
===========================
- src/Kernel/Factories/RecurrenceConfigFactory.php  -12
- src/Kernel/ValueObjects/Configuration/RecurrenceConfig.php  -24
         

Complexity increasing per file
==============================
- src/Kernel/Factories/RecurrenceConfigFactory.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mundipagg/ecommerce-module-core/pullRequest?prid=4985325&bid=16461689)

**Comment by wallaceSF** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 11
           

Complexity increasing per file
==============================
- src/Recurrence/Services/Rules/CompatibleRecurrenceProducts.php  3
- src/Recurrence/Services/Rules/ProductListInCart.php  1
- src/Recurrence/Services/Rules/NormalWithRecurrenceProduct.php  3
- src/Kernel/Factories/RecurrenceConfigFactory.php  3
- src/Recurrence/Services/Rules/MoreThanOneRecurrenceProduct.php  2
- src/Recurrence/Services/Rules/CurrentProduct.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mundipagg/ecommerce-module-core/pullRequest?prid=4985325&bid=16461689)

**Comment by wallaceSF** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 16
           

Complexity increasing per file
==============================
- tests/Recurrence/Services/Rules/ProductListInCartTest.php  1
- tests/Recurrence/Services/Rules/CompatibleRecurrenceProductsTest.php  1
- src/Recurrence/Services/Rules/MoreThanOneRecurrenceProduct.php  2
- src/Recurrence/Services/Rules/CompatibleRecurrenceProducts.php  3
- tests/Recurrence/Services/Rules/CurrentProductTest.php  1
- src/Recurrence/Services/Rules/CurrentProduct.php  1
- tests/Recurrence/Services/Rules/NormalWithRecurrenceProductTest.php  2
- src/Recurrence/Services/Rules/NormalWithRecurrenceProduct.php  3
- src/Recurrence/Services/Rules/ProductListInCart.php  1
- tests/Recurrence/Services/Rules/MoreThanOneRecurrenceProductTest.php  2
- src/Kernel/Factories/RecurrenceConfigFactory.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mundipagg/ecommerce-module-core/pullRequest?prid=4985325&bid=16461689)

File: src/Recurrence/Services/Rules/NormalWithRecurrenceProduct.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+        return;
+    }
+
+    public function getError()
+    {
+        return $this->error;
+    }
+
+    public function setError($error)
```
Comment (wallaceSF) [inline src/Recurrence/Services/Rules/NormalWithRecurrenceProduct.php]: Esse método está como público. Pode mudar para uma classe abstrata sem corpo protected ou remover esse contrato dá interface.

File: src/Kernel/I18N/PTBR.php [hunk 2]
```diff
@@ -77,6 +77,7 @@ protected function getTable()
             'New order status: %s' => 'Novo status do pedido: %s',
             'Subscription invoice created: %.2f' => 'Fatura de assinatura criada: %.2f',
             'Url boleto' => 'Boleto url',
+            'The recurrence products should have the same payment configuration and the same interval' => 'Os produtos recorrentes devem ter a mesma configuração de pagamento e intervalo',
```
Comment (michelpl) [inline src/Kernel/I18N/PTBR.php]: Só um comentário: O cliente final não sabe o que é um produto recorrente, então quando formos colocar mais mensagens desse tipo, pensar em algo que o cliente entenda.

File: src/Recurrence/Services/Rules/CompatibleRecurrenceProducts.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
+<?php
+
```
Comment (michelpl) [inline src/Recurrence/Services/Rules/CompatibleRecurrenceProducts.php]: Trocar o nome da pasta Rules pra CartRules

**Review by michelpl — CHANGES_REQUESTED** [review]

**Comment by wallaceSF** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 16
           

Complexity increasing per file
==============================
- tests/Recurrence/Services/CartRules/CurrentProductTest.php  1
- src/Recurrence/Services/CartRules/CurrentProduct.php  1
- src/Recurrence/Services/CartRules/ProductListInCart.php  1
- tests/Recurrence/Services/CartRules/CompatibleRecurrenceProductsTest.php  1
- tests/Recurrence/Services/CartRules/ProductListInCartTest.php  1
- tests/Recurrence/Services/CartRules/NormalWithRecurrenceProductTest.php  2
- src/Kernel/Factories/RecurrenceConfigFactory.php  3
- tests/Recurrence/Services/CartRules/MoreThanOneRecurrenceProductTest.php  2
- src/Recurrence/Services/CartRules/MoreThanOneRecurrenceProduct.php  2
- src/Recurrence/Services/CartRules/CompatibleRecurrenceProducts.php  3
- src/Recurrence/Services/CartRules/NormalWithRecurrenceProduct.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mundipagg/ecommerce-module-core/pullRequest?prid=4985325&bid=16461689)

**Comment by wallaceSF** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 16
           

Complexity increasing per file
==============================
- src/Recurrence/Services/CartRules/CompatibleRecurrenceProducts.php  3
- tests/Recurrence/Services/CartRules/CompatibleRecurrenceProductsTest.php  1
- src/Recurrence/Services/CartRules/ProductListInCart.php  1
- tests/Recurrence/Services/CartRules/CurrentProductTest.php  1
- tests/Recurrence/Services/CartRules/MoreThanOneRecurrenceProductTest.php  2
- src/Recurrence/Services/CartRules/MoreThanOneRecurrenceProduct.php  2
- tests/Recurrence/Services/CartRules/ProductListInCartTest.php  1
- src/Recurrence/Services/CartRules/CurrentProduct.php  1
- src/Recurrence/Services/CartRules/NormalWithRecurrenceProduct.php  3
- src/Kernel/Factories/RecurrenceConfigFactory.php  3
- tests/Recurrence/Services/CartRules/NormalWithRecurrenceProductTest.php  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mundipagg/ecommerce-module-core/pullRequest?prid=4985325&bid=16461689)

**Review by michelpl — APPROVED** [review]

**Review by wallaceSF — APPROVED** [review]

**Pull request merged**